hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/encryption/CSEUtils.java

```diff
@@ -0,0 +1,214 @@
+    return CSE_KMS.getMethod().equals(encryptionMethod) ||
+        CSE_CUSTOM.getMethod().equals(encryptionMethod);
+  }
+
+  /**
+   * Checks if a given S3 object is encrypted or not by checking following two conditions
+   * 1. if object metadata contains x-amz-cek-alg
+   * 2. if instruction file is present
+   *
+   * @param s3Client S3 client
+   * @param factory   S3 request factory
+   * @param key      key value of the s3 object
+   * @return true if S3 object is encrypted
+   */
+  public static boolean isObjectEncrypted(S3Client s3Client, RequestFactory factory, String key) {
+    HeadObjectRequest.Builder requestBuilder = factory.newHeadObjectRequestBuilder(key);
+    HeadObjectResponse headObjectResponse = s3Client.headObject(requestBuilder.build());
+    if (headObjectResponse.hasMetadata() &&
+        headObjectResponse.metadata().get(CRYPTO_CEK_ALGORITHM) != null) {
+      return true;
+    }
+    HeadObjectRequest.Builder instructionFileRequestBuilder =
+        factory.newHeadObjectRequestBuilder(key + S3_ENCRYPTION_CSE_INSTRUCTION_FILE_SUFFIX);
+    try {
+      s3Client.headObject(instructionFileRequestBuilder.build());
+      return true;
+    } catch (NoSuchKeyException e) {
+      // Ignore. This indicates no instruction file is present
+    }
+    return false;
+  }
+
+  /**
+   * Get the unencrypted object length by either subtracting
+   * {@link InternalConstants#CSE_PADDING_LENGTH} from the object size or calculating the
+   * actual size by doing S3 ranged GET operation.
+   *
+   * @param s3Client           S3 client
+   * @param bucket             bucket name of the s3 object
+   * @param key                key value of the s3 object
+   * @param factory            S3 request factory
+   * @param contentLength      S3 object length
+   * @param headObjectResponse response from headObject call
+   * @param cseRangedGetEnabled is ranged get enabled
+   * @param cseReadUnencryptedObjects is reading of une
+   * @return unencrypted length of the object
+   * @throws IOException IO failures
+   */
+  public static long getUnencryptedObjectLength(S3Client s3Client,
+      String bucket,
+      String key,
+      RequestFactory factory,
+      long contentLength,
+      HeadObjectResponse headObjectResponse,
+      boolean cseRangedGetEnabled,
+      boolean cseReadUnencryptedObjects) throws IOException {
+
+    if (cseReadUnencryptedObjects) {
+      // if object is unencrypted, return the actual size
+      if (!isObjectEncrypted(s3Client, factory, key)) {
+        return contentLength;
+      }
+    }
+
+    // check if unencrypted content length metadata is present or not.
+    if (headObjectResponse != null) {
+      String plaintextLength = headObjectResponse.metadata().get(UNENCRYPTED_CONTENT_LENGTH);
+      if (headObjectResponse.hasMetadata() && plaintextLength !=null &&
+          !plaintextLength.isEmpty()) {
+        return Long.parseLong(plaintextLength);
+      }
+    }
+
+    if (cseRangedGetEnabled) {
```

Comment:
I think previously we always just subtracted 16, was that incorrect?

Reply:
V2 and V3 always pads 16 bytes so it makes sense to subtract 16 bytes but to have compatibility with v1 client which pads bytes in multiple of 16 simply subtracting 16 bytes leads to wrong result/exception.